runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/graph/GreedyStageFuser.java

```diff
@@ -125,6 +128,7 @@ public static ExecutableStage forGrpcPortRead(
     return ImmutableExecutableStage.of(
         environment,
         inputPCollection,
+        ImmutableSet.copyOf(sideInputs),
```

Comment:
It's a bit weird to pessimistically copy fields here when we know that ImmutableExecutableStage.of() already does this.

Reply:
Point. The deep immutability of `ImmutableExecutableStage`, or at least it's collections, is a responsibility of that class, so it's really being defensive against abstraction failure.